controllers/UserController.php

```diff
@@ -197,4 +197,41 @@ protected function findModel($id)
         }
         throw new NotFoundHttpException('The requested page does not exist.');
     }
+
+    public function actionRequestEmailVerification()
+    {
+        /** @var User $user */
+        $user = Yii::$app->user->identity;
+
+        if ($user->email_verified) {
+            Yii::$app->getSession()->setFlash('success', 'Your email is already verified.');
+        } else {
+            $this->sendEmailVerificationEmail($user);
+            Yii::$app->getSession()->setFlash('success', 'Please check your email to verify it.');
+        }
+
+        return Yii::$app->user->isGuest ? $this->goHome() : $this->redirect(['user/profile']);
+    }
+
+    private function sendEmailVerificationEmail(User $user)
+    {
+        if ($user) {
+            if (!User::isEmailVerificationTokenValid($user->email_verification_token)) {
+                $user->generateEmailVerificationToken();
```

Comment:
this method should do save() internally, makes code below easier for not checking return value of save().

Reply:
Internally where?